R/generate.R

```diff
@@ -0,0 +1,105 @@
+  names(args) <- table$column_name
+  cols <- lapply(table$column_name, as.name)
+  names(cols) <- table$column_name
+  expr <- as.call(c(list(quote(data.frame)),
+                    cols,
+                    list(stringsAsFactors = FALSE)))
+
+  body <- c(
+    Map(function(col_name, is_nullable, r_type) {
+      env <- list(x = as.name(col_name), y = r_type, z = is_nullable == "YES")
+      substitute_(quote(validate(x, y, z)), env)
+    }, table$column_name, table$is_nullable, table$r_type),
+    list(expr))
+
+  as_function(args, body, env)
+}
+
+build_args <- function(is_nullable, r_type, ...) {
+  res <- rep(alist(. =), length(is_nullable))
+  res[is_nullable == "YES" & r_type == "integer"] <- alist(. = NA_integer_)
+  res[is_nullable == "YES" & r_type == "numeric"] <- alist(. = NA_real_)
+  res[is_nullable == "YES" & r_type == "character"] <- alist(. = NA_character_)
+  res[is_nullable == "YES" & r_type == "logical"] <- alist(. = NA)
+  res[is_nullable == "YES" & r_type == "POSIXct"] <- alist(. = NA_real_)
+  res
+}
+
+substitute_ <- function(expr, env) {
+  eval(substitute(substitute(y, env), list(y = expr)))
+}
+
+as_function <- function(args, body, env) {
+  body_expr <- list(as.call(c(list(as.name("{")), unname(body))))
+  as.function(c(args, body_expr), env)
+}
+
+tables_with_types <- function(tables) {
+  tables$r_type <- "character"
+  tables[tables$data_type %in% int_types, "r_type"] <- "integer"
+  tables[tables$data_type %in% logi_types, "r_type"] <- "logical"
+  tables[tables$data_type %in% num_types, "r_type"] <- "double"
+  tables[tables$data_type %in% date_types, "r_type"] <- "POSIXct"
+  split(tables, tables$table_name)
+}
+
+fetch_tables <- function(con, schema_name) {
+  safe_param <- DBI::dbQuoteLiteral(con, schema_name)
+  query <- paste0("SELECT column_name, table_name, data_type, column_default, is_nullable FROM
+  information_schema.columns WHERE table_schema = ", safe_param)
+  DBI::dbGetQuery(con, query)
+}
+
+#' Generate mocking functions for all tables in the given db schema
+#'
+#' @param con DBI connection
+#' @param schema_name name of the db schema
+#' @returns list of named functions corresponding to the tables in the schema
+#' @export
+generate <- function(con, schema_name = "public") {
+  dbname <- DBI::dbGetInfo(con)$dbname
+  dest <- file.path("inst/fakerbase", dbname, schema_name)
+  dir.create(dest, recursive = TRUE, showWarnings = FALSE)
+  tables <- tables_with_types(tables = fetch_tables(con, schema_name))
+  fns <- lapply(tables, build)
+  saveRDS(fns, file.path(dest, "generated.rds"))
+  fns
+}
+
+#' Load previously generated mocking functions for all tables in the given db and schema
+#'
+#' @param dbname name of the database
+#' @param schema_name name of the db schema
+#' @returns list of named functions corresponding to the tables in the schema
+#' @export
+load <- function(dbname, schema_name = "public") {
+  db_dir <- file.path("inst/fakerbase", dbname)
```

Comment:
Sorry should have noticed this before.

I think what this should say is something like:

```r
db_dir <- system.file(file.path("fakerbase", dbname), package = package, mustWork = TRUE)
```

with this function also accepting `package` as an arg.  So in vimpact we might use

```
fakerbase::load("montagu", "vimpact")
```

to load generated files.  This will produce a fairly terrible error message if the user has not generated files though, but we can tidy that up later

Reply:
Why require that the consumer be a package?

Reply:
you're sort of implying that that with the `inst/` writing atm